### PR TITLE
Various const eval cleanups

### PR DESCRIPTION
**Pull request by oli-obk**
This pulls out the pure refactorings from https://github.com/rust-lang/rust/pull/116564

r? @RalfJung

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 1]
```diff
@@ -1069,23 +1070,33 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+        let read_provenance = matches!(
+            layout.abi,
+            Abi::Scalar(abi::Scalar::Initialized { value: abi::Primitive::Pointer(..), .. })
+        );
+
+        let scalar = alloc
+            .inner()
+            .read_scalar(self, alloc_range(Size::ZERO, layout.size), read_provenance)
+            .map_err(|err| err.to_interp_error(alloc_id))?;
+        self.adjust_const_scalar(scalar)
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:1099]: As before I'm not happy with bypassing all our nice abstractions here. Why is this necessary?

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 2]
```diff
@@ -939,6 +942,64 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         self.validate_operand_internal(op, path, Some(ref_tracking), Some(ctfe_mode))
     }
 
+    #[inline(always)]
+    pub fn const_validate_mplace(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/validity.rs:946]: I was quite happy to have this very const-specific code *not* in the `interpret` file. Why is it being moved?

Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/validity.rs:946]: You mean like `const_validate_operand` 😆 

I can move it into the const_eval module, all I wanted was to make it a method

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/validity.rs:946]: Well the names are rather misleading, this one here is doing the full handling of recursive validation and error reporting.

File: compiler/rustc_const_eval/src/interpret/eval_context.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:1099]: I got rid of it (see the last commit for how it works now)

File: compiler/rustc_const_eval/src/interpret/validity.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/validity.rs]: I moved it back to `const_eval.rs`

**Comment by oli-obk** [comment]
This is ready for review again

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:1]: This is just outlining, right? These new functions only have a single call site?

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 3]
```diff
@@ -1074,17 +1074,14 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
         let gid = GlobalId { instance, promoted: None };
-        // For statics we pick `ParamEnv::reveal_all`, because statics don't have generics
-        // and thus don't care about the parameter environment. While we could just use
-        // `self.param_env`, that would mean we invoke the query to evaluate the static
-        // with different parameter environments, thus causing the static to be evaluated
-        // multiple times.
-        let param_env = if self.tcx.is_static(gid.instance.def_id()) {
-            ty::ParamEnv::reveal_all()
+        let val = if self.tcx.is_static(gid.instance.def_id()) {
+            let alloc_id = self.tcx.reserve_and_set_static_alloc(gid.instance.def_id());
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:1078]: Ah, clever!
Could you a comment explaining why we're doing it this way?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:1078]: In particular, this doesn't actually add the assertion that `eval_to_allocation_raw` is never called on statics, or does it?

Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:1078]: No, because that isn't true in this PR yet.

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs
Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:1]: yes

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 4]
```diff
@@ -336,13 +336,11 @@ pub fn eval_to_allocation_raw_provider<'tcx>(
...
-            if let Err(error) = validation {
-                Err(const_report_error(&ecx, error, alloc_id))
-            } else {
+            validation
+                // Validation failed, report an error.
+                .map_err(|error| const_report_error(&ecx, error, alloc_id))
                 // Convert to raw constant
-                Ok(ConstAlloc { alloc_id, ty: mplace.layout.ty })
-            }
+                .map(|()| ConstAlloc { alloc_id, ty: mplace.layout.ty })
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:350]: Honestly I find the `match` more readable than using higher-order functions... but I don't care very strongly.

Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:350]: I force pushed away the commit that did this

**Comment by RalfJung** [comment]
Thanks!
@bors r+
